Source/WTF/wtf/AbstractThreadSafeRefCountedAndCanMakeWeakPtr.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2022-2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/AbstractRefCounted.h>
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WTF {
+
+class AbstractThreadSafeRefCountedAndCanMakeWeakPtr : public AbstractRefCounted {
+public:
+    virtual ThreadSafeWeakPtrControlBlock& controlBlock() const = 0;
+
+private:
+    template<typename> friend class ThreadSafeWeakHashSet;
+    virtual size_t weakRefCount() const = 0;
+};
+
+// Convinience macro that implements AbstractThreadSafeRefCountedAndCanMakeWeakPtr for you.
+#define WTF_ABSTRACT_THREAD_SAFE_REF_COUNTED_AND_CAN_MAKE_WEAK_PTR_IMPL(Class, ... /* DestructionThread */) \
+    void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Class, ##__VA_ARGS__>::ref(); } \
```

Comment:
I really think &lt;Class, ##__VA_ARGS__&gt; is unnecessary, so you don't need any parameters for this macro.  If that doesn't compile for some reason, this is fine.

Reply:
Sure, I'll try it.